Archlinux support

chr0mag opened the pull request:
---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [X] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [X] Is the subject and message clear and concise?
- [X] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?

bmr-cymru commented:
Please see my comments in the issue you opened - the policy needs to also define a new tagging class (or accept one or more of the existing ones), and all your patches need a clear subject (and optional body description unless they are very simple).

bmr-cymru commented on sos/plugins/pacman.py:
```diff
@@ -0,0 +1,42 @@
+
+        self.add_cmd_output(
+            "ls -l /usr/share/libalpm/hooks /etc/pacman.d/hooks",
+            suggest_filename="hooks")
```
Closing paren should be on the following line, indented two levels (like the one on line 40).

chr0mag replied:
Done. I am running `pycodestyle` as a git pre-commit hook but it doesn't seem to catch this.

bmr-cymru commented on sos/policies/arch.py:
```diff
@@ -0,0 +1,53 @@
+import os
```
What's this for?

chr0mag replied:
It's unused, and has been removed. (Incidentally, `pyflakes` shows 17 other instances of unused imports under `../sos`.)

bmr-cymru commented on sos/policies/arch.py:
```diff
@@ -0,0 +1,53 @@
+import os
+from sos.policies import PackageManager, LinuxPolicy
+from sos.plugins import Plugin
```
Need to import either a newly defined tagging class or one of more of the existing ones.

chr0mag commented:
@bmr-cymru - Thanks for taking the time to review this. I've commited a couple of changes (with more useful, descriptive messages) based on your feedback. A new ArchPlugin tagging class now exists and is used by the Arch policy and Pacman plugin. For the moment, I've only tagged the Pacman plugin with the new tagging class, so its the only one executed on an Arch system. I will work my way through tagging existing plugins relevant to Arch if the current changes look reasonable to you.

chr0mag commented:
I've tagged 37 classes with the new ArchPlugin tagging class. Including the pacman plugin, this makes 38 plugins tagged with this class. No other code changes were made to the plugins apart from importing ArchPlugin and appending this to the list of tagging classes.

All 37 plugins produce useful output as is on an Arch system. A sample sosreport is here:
[sosreport-38-plugins-tagged-20180214213345.tar.gz](https://github.com/sosreport/sos/files/1726579/sosreport-38-plugins-tagged-20180214213345.tar.gz)

pmoravec commented on sos/plugins/ipa.py:
```diff
@@ -59,7 +59,7 @@ def retrieve_pki_logs(self, ipa_version):
                "/var/log/pki/pki-tomcat/ca/transactions",
                "/var/log/pki/pki-tomcat/catalina.*",
-               "/var/log/pki/pki-ca-spawn.*"
+               "/var/log/pki/pki-ca-spawn.*",
```
This typo has been recently fixed.

Also in genral it makes sense to have one logical change (adding ArchPlugin to many plugins, e.g.) in one commit and an independent fix like this one in another.

pmoravec commented on sos/plugins/openvswitch.py:
```diff
@@ -135,6 +135,18 @@ def setup(self):
                             "ovs-ofctl -O %s dump-ports-desc %s" % (flow, br)
                         ])
 
+        # Gather info on the DPDK mempools associated with each DPDK port
```
Also recently merged / fixed.

pmoravec commented on sos/plugins/rabbitmq.py:
```diff
@@ -28,9 +28,26 @@ class RabbitMQ(Plugin, RedHatPlugin, DebianPlugin, UbuntuPlugin):
-        self.add_cmd_output("rabbitmqctl report")
-        self.add_cmd_output("rabbitmqctl cluster_status")
-        self.add_cmd_output("rabbitmqctl list_policies")
+        container_status = self.get_command_output(
```
Same here.

pmoravec commented:
No more plugins are useful / applicable to ArchLinux?

I spotted 3 places of code change that is already in upstream (ipa, openvswicth, rabbitmq plugins), otherwise OK from me.

chr0mag commented:
@pmoravec - There are many more plugins that are applicable to ArchLinux. The 37 tagged here were the low hanging fruit that a) didn't require anything other than tagging and b) packagesrequired by the plugins were already installed on a system I had access to in order to test things worked. There are several dozen additional plugins that will likely work as they are and just need to be tagged, but I have yet to test these and didn't want to throw them over the fence without doing so. On top of that there are at least a dozen plugins where file path and/or package names are different in Arch so code changes are required before they'll work.

bmr-cymru commented:
Will try to get this re-reviewed and included for 3.6: it would be nice to be able to declare Arch officially supported.

TurboTurtle commented:
This missed 3.6, moving out to 3.7

chr0mag commented:
I rebased against master on July 11th. I dropped a bad commit [6f5f093](https://github.com/sosreport/sos/pull/1198/commits/6f5f09323f7d9eb5c3497729e8d5cbe45f0b7a3e) as well as the corresponding revert commit [536527d](https://github.com/sosreport/sos/pull/1198/commits/536527d9f1eb230714ca972fefd0be5a079e8a98).

battlemidget commented:
Can we resurrect this for next release? Would love to see another supported distribution

pmoravec commented:
Before this can be merged:

- branch conflicts must be resolved
- pleaseremove the "[distupgrade] Fixes #1197." and its revert commit
- I see Bryn's comments reacted with "done/applied", but the current commits in the branch still show the original work (e.g.with `suggest_filename="hooks")` line or reduntant `import os` etc.)

pmoravec commented:
Hello,
some feedback to the review?

chr0mag commented:
Hello,

- branch has been rebased onto master and conflicts resolved
- bad commit and it's revert were dropped during rebase
- I don't see the redundant `import os` or the closing paren issue in the branch. Those issues were addressed in https://github.com/sosreport/sos/pull/1198/commits/57c25a4f0a4c643267cce66cfc6ed1ea23835167 .

pmoravec commented on sos/policies/arch.py:
```diff
@@ -0,0 +1,48 @@
+    valid_subclasses = [ArchPlugin]
+
+    def __init__(self, sysroot=None):
+        super(LinuxPolicy, self).__init__(sysroot=sysroot)
```
there should be `super(ArchPolicy, self)....` to call `__init__` method of parent class of *this* `ArchPolicy` class (i.e. to call the method of `LinuxPolicy` - not parent class of `LinuxPolicy` that is class `Policy` ;-)

pmoravec commented on sos/plugins/devicemapper.py:
```diff
@@ -6,10 +6,15 @@
+                   RedHatPlugin,
+                   DebianPlugin,
+                   UbuntuPlugin,
+                   ArchPlugin):
```
Maybe just personal stylistic preference, but listing plugins like https://github.com/sosreport/sos/blob/master/sos/plugins/kata_containers.py#L13-L14 seems better to me than scattere one plugin per one line. But as I wrote, this is rather a personal preference only.

pmoravec commented:
Apart of the one or two comments in code, some commits can be squashed together, like "Fixed bare except." or "[pacman] add pacman.log to output" - I dont have strong opinion how strict we should be on git history with such redundant/extra commits.

TurboTurtle commented on sos/policies/arch.py:
```diff
@@ -0,0 +1,48 @@
+from sos.policies import PackageManager, LinuxPolicy
```
This needs the copyright header that is present in all other sos files:

```
# This file is part of the sos project: https://github.com/sosreport/sos
#
# This copyrighted material is made available to anyone wishing to use,
# modify, copy, or redistribute it subject to the terms and conditions of
# version 2 of the GNU General Public License.
#
# See the LICENSE file in the source distribution for further information.
```


Optionally, add a copyright line above it if you want to. E.G. `# Copyright (C) 2019 Optional Company, Your Name <your email>
`

chr0mag commented:
Thanks for the feedback. Call to 'super' and formatting have been addressed. Copyright has been added and minor commits squashed.

pmoravec commented on sos/policies/arch.py:
```diff
@@ -0,0 +1,56 @@
+
+class Pacman(PackageManager):
+
+    def get_pkg_list(self):
```
Rather a curious question than a comment to change anything: why `get_pkg_list` from parent cant be used? Because it expects packages as strings in format like:

    avalon-framework|4.3|10.el7

while `pacman --query` returns a list of lines like:

    vlc 2.2.2-3

?

So then replacing ' ' by '|' can help in using the parent class method? (then it is imho worth extending the parent class method to use user-defined separator, like:

    def get_pkg_list(self, sep='|'):
        """Returns a dictionary of packages in the following
        format::

            {'package_name': {'name': 'package_name',
                              'version': 'major.minor.version'}}

        """
        if self.query_command:
            cmd = self.query_command
            pkg_list = shell_out(
                cmd, timeout=0, chroot=self.chroot
            ).splitlines()

            for pkg in pkg_list:
                if sep not in pkg:
                    continue
                elif pkg.count(sep) == 1:
                    name, version = pkg.split(sep)
                    release = None
                elif pkg.count(sep) == 2:
                    name, version, release = pkg.split(sep)
                self.packages[name] = {
                    'name': name,
                    'version': version.split(".")
                }
                release = release if release else None
                self.packages[name]['release'] = release

        return self.packages

Would this work?

chr0mag replied:
Yes, that's correct. `pacman --query` output is space-delimited and there is no output formatting switch to change this like some other package managers. And yes, being able to specify the delimiter would work. Is that something you'd like to see as part of this PR?

pmoravec replied:
I think extending `get_pkg_list` by separator is better option than copying majority of its code. Any thoughts, @bmr-cymru or @TurboTurtle ?

pmoravec commented:
Apart of the one comment / question, +1 for the PR. Thanks for the work done.

TurboTurtle commented:
Cycling back around on this, @chr0mag are you still interested in adding Arch support to sos?

I think the only outstanding bit at this point is the separator extension for `get_pkg_list()`.

And a rebase against the current master of course ;-)

TurboTurtle commented:
One more check-in attempt. @chr0mag are you still interested in bringing Arch support to sos?

TurboTurtle commented:
Closing due to inactivity. If you would like to bring Arch support to sos, please open a new PR against current master.

hab-pf commented:
fixed it up for current master here: https://github.com/sosreport/sos/compare/main...hab-pf:archlinux-support

TurboTurtle commented:
> fixed it up for current `main` here: [main...hab-pf:archlinux-support](https://github.com/sosreport/sos/compare/main...hab-pf:archlinux-support)

@hab-pf please open up a new PR for this. Most of that looks fine, except the new `Pacman` package manager will need a little refinement:

1. Needs to be placed in (newly created) `sos/policies/package_managers/pacman.py` rather than in `arch.py`.
2. Need to define the `query_command` and `verify_command` class attrs. If `pacman --query` supports a formatting option, I'd strongly recommend using that to use the default `get_pkg_list()` like we do for both `rpm` and `dpkg`.

hab-pf commented:
I don't really understand the changes; I only brought the branch up to date. I would not be the best person to finish this. I can if nobody else steps up.
